packages/grid/data-grid/src/tests/components.DataGrid.test.tsx

```diff
@@ -49,6 +50,70 @@ describe('<DataGrid /> - Components', () => {
     });
   });
 
+  describe('componentsProps', () => {
+    it('should pass the props from componentsProps.cell to the cell', () => {
+      render(
+        <div style={{ width: 300, height: 500 }}>
+          <DataGrid
+            {...baselineProps}
+            hideFooter
+            disableVirtualization
+            componentsProps={{ cell: { 'data-name': 'foobar' } }}
+          />
+        </div>,
+      );
+      expect(getCell(0, 0)).to.have.attr('data-name', 'foobar');
+    });
+
+    it('should pass the props from componentsProps.row to the row', () => {
+      render(
+        <div style={{ width: 300, height: 500 }}>
+          <DataGrid
+            {...baselineProps}
+            hideFooter
+            disableVirtualization
+            componentsProps={{ row: { 'data-name': 'foobar' } }}
+          />
+        </div>,
+      );
+      expect(getRow(0)).to.have.attr('data-name', 'foobar');
+    });
+  });
+
+  describe('components', () => {
+    it('should render the cell with the component given in components.Cell', () => {
+      render(
+        <div style={{ width: 300, height: 500 }}>
+          <DataGrid
+            {...baselineProps}
+            hideFooter
+            disableVirtualization
+            components={{
+              Cell: ({ rowIndex, colIndex }) => (
+                <span role="cell" data-rowindex={rowIndex} data-colindex={colIndex} />
+              ),
+            }}
+          />
+        </div>,
+      );
+      expect(getCell(0, 0).tagName).to.equal('SPAN');
+    });
+
+    it('should render the row with the component given in components.Row', () => {
+      render(
+        <div style={{ width: 300, height: 500 }}>
+          <DataGrid
+            {...baselineProps}
+            hideFooter
+            disableVirtualization
+            components={{ Row: ({ rowIndex }) => <span role="row" data-rowindex={rowIndex} /> }}
```

Comment:
Do we agree that if somebody overrides `row` or `cell`, he must to copy / paste the code in`GridRow`, otherwise the basic event listeners won't be created for instance.

This mean that every modification we do on this component is technically a breaking change.
I think it makes the API very fragile. Someone overriding this kind of component risks to have broken behavior a soon as we add new features that requires to add code to those components.

What is our approach on this subject ?

- the `components` props is for advanced application and those who use it must be aware that this part may not respect semver ?
- we respect semver on those components and will never add any required code to them between major (don't think it's viable)
- 3rd option ?

Reply:
> Do we agree that if somebody overrides row or cell, he must to copy / paste the code inGridRow, otherwise the basic event listeners won't be created for instance.

Yes, that's the purpose of the prop. It's for extreme use cases, when the row or cell needs to be completely overridden. For simple cases, using `componentsProps` is sufficient.

> This mean that every modification we do on this component is technically a breaking change.

Yes, but the row and cell are quite stable. We didn't made big changes on both in the past.

> What is our approach on this subject ?

Option 2 maybe. These components are dumb and don't have any complex functionality, everything comes via props. I don't see a feature which will require breaking changes on them, besides the virtualization.

Reply:
> We have to modify these components if we want to add a class to the DOM, as we did when we introducted the edition mode for instance.

In this case, adding a class is not a breaking change.

> But if you think that these components are stable enough, I just think that it would be nice to add a clear warning here https://mui.com/components/data-grid/components/#components, to avoid complains due to complex migrations (even if they only come in the next major).

That can be done latter once we decide what to do after the stable release. For now, any breaking change to the row or cell component should be mentioned.

Reply:
Indded
Adding a `data-attr` that we use in the rest of the codebase like `data-id` would.


> That can be done latter once we decide what to do after the stable release. For now, any breaking change to the row or cell component should be mentioned.

OK :+1: